File: BaekJoon/Java/6603_로또_DFS_BackTracking_Math.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

// 로또 6603 DFS, BackTracking, Math
// O( kC6 )


public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = null;
        StringBuilder sb = new StringBuilder();

        while (true) {
            st = new StringTokenizer(br.readLine());

            int k = Integer.parseInt(st.nextToken());

            if (k == 0) break;

            int[] s = new int[k];
            for (int i = 0; i < k; ++i) {
                s[i] = Integer.parseInt(st.nextToken());
            }

            DFS(s, new int[6], 0, 0, sb);
            sb.append("\n");
        }

        System.out.println(sb.toString());
    }

    private static void DFS(int[] s, int[] cur, int start, int depth, StringBuilder sb) {
        if (depth == 6) {
            for (int n : cur) {
                sb.append(n).append(" ");
            }
            sb.append("\n");
            return;
        }

        for (int i = start; i < s.length; ++i) {
            cur[depth] = s[i];
            DFS(s, cur, i + 1, depth + 1, sb);
        }
    }
}

File: BaekJoon/6603_로또_DFS_BackTracking_Math.cpp
#include <iostream>
#include <vector>

using namespace std;

// 로또 6603 DFS, BackTracking, Math
// O( kC6 )


void DFS(const vector<int>& s, vector<int>& cur, int start, int depth)
{
    // 6개 숫자를 모두 선택했다면 출력
    if (depth == 6)
    {
        for (int i = 0; i < 6; ++i)
        {
            cout << cur[i] << " ";
        }
        cout << "\n";
        return;
    }

    // 현재 위치부터 가능한 모든 숫자 재귀 선택
    for (int i = start; i <= s.size() - 6 + depth; ++i)
    {
        cur[depth] = s[i];
        DFS(s, cur, i + 1, depth + 1);
    }
}

int main(void)
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    
    int k;
    while (cin >> k && k != 0)
    {
        vector<int> s(k);
        for (int& e : s)
        {
            cin >> e;
        }
        
        vector<int> cur(6); // 6개 로또 벡터
        
        DFS(s, cur, 0, 0);

        cout << "\n";
    }


    return 0;
}
